src/PooledArrays.jl

```diff
@@ -448,7 +468,8 @@ Base.convert(::Type{Array}, pa::PooledArray{T, R, N}) where {T, R, N} = convert(
 
 Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Int)
     idx = DataAPI.refarray(A)[I]
-    iszero(idx) && throw(UndefRefError())
+    # if eltype(A) does not allow Missing then also 1 is an error
+    idx <= !(eltype(A) <: Missing) && throw(UndefRefError())
```

Comment:
Hmmm, I think something like the following should be faster (as it avoids the core subtyping algorithm):
```julia
@assert Base.nonmissingtype(T) === T || !isone(idx)
```
I think I like doing an `@assert` here instead since if `idx == 1`, then it means something is corrupt, not `#undef`.

Reply:
We could do this also. I wanted to squeeze it into one check (to reduce the cost of `getindex`).

However, first we should decide if we go for the design I proposed here or the design similar to what CategoricalArrays.jl does (as the other approach does not have this issue).

Reply:
The other approach is having `0` ref be `missing`, rigth?

Reply:
Right. The issue is the following:
```
julia> x = categorical(["a", "b"])
2-element CategoricalArray{String,1,UInt32}:
 "a"
 "b"

julia> y = categorical(["a", missing])
2-element CategoricalArray{Union{Missing, String},1,UInt32}:
 "a"
 missing

julia> xs = similar(x)
2-element CategoricalArray{String,1,UInt32}:
 #undef
 #undef

julia> ys = similar(y)
2-element CategoricalArray{Union{Missing, String},1,UInt32}:
 missing
 missing
```
so this would be breaking (i.e. it is impossible to have `#undef` in `CategoricalArray` if its eltype allows `Missing`).

Reply:
and then you have:
```
julia> DataAPI.refpool(ys)
1-element CategoricalArrays.CategoricalRefPool{Union{Missing, CategoricalValue{String, UInt32}}, CategoricalPool{String, UInt32, CategoricalValue{String, UInt32}}} with indices 0:0:
 missing
```
which means one needs a custom type to handle non 1-based indexing.

Reply:
Hmmmm, I don't think that's awful behavior, but indeed a bit inconsistent. I mean, not letting users get `#undef` isn't the worst, IMO, but it does feel like cheating a bit.

Reply:
I feel like we'd have a similar problem with `DataAPI.refpool` on PooledArray w/ no `missing` values, right? Because `pool[1]` would be `missing`, even though the PooledArray doesn't allow `missing`s?

Reply:
> Because `pool[1]` would be missing, even though the PooledArray doesn't allow `missing`s?

The option is to have a custom `refpool` type (like CategoricalArrays.jl)